File: python3/decoded_string_at_index.py
"""
Space   : O(1)
Time    : O(n)
"""


class Solution:
    def decodeAtIndex(self, S: str, K: int) -> str:
        size = 0
        n = len(S)

        for i in range(n):
            if S[i].isdigit():
                size *= int(S[i])
            else:
                size += 1

        for j in range(n-1, -1, -1):
            K %= size

            if K == 0 and not S[j].isdigit():
                if K == 0:
                    return S[j]
                
                size -= 1
            else:
                size //= int(S[j])

        return S[K]

File: golang/decoded_string_at_index.go
package main

import (
	"fmt"
	"strconv"
)

/*
Space	: O(1)
Time	: O(n)
*/

func decodeAtIndex(s string, k int) string {
	size := 0
	n := len(s)

	for i := 0; i < n; i++ {
		if dig, err := strconv.Atoi(string(s[i])); err == nil {
			size *= dig
		} else {
			size++
		}
	}

	fmt.Println(size)
	for j := n - 1; j >= 0; j-- {
		k %= size

		dig, err := strconv.Atoi(string(s[j]))

		if err != nil {
			if k == 0 {
				return string(s[j])
			}
			size--
		} else {
			size /= dig
		}
	}

	return string(s[k])
}
